### ISSUE TITLE
support fetchall with executemany, ORM integration

### ISSUE THREAD

**Issue by kataev** [sqlalchemy/sqlalchemy#5263]
Hello, i'm just asking about why `bulk_save_objects` and `bulk_insert_mappings` with `return_defaults=True` don't use `executemany` at least on pg?
Why we can't use one query for insert many rows in orm and when set id for this instances?
PostgreSQL don't preserve order in result returnings? 
```python
import logging

import sqlalchemy as sa
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session

Base = declarative_base('postgresql://postgres:postgres@localhost/postgres')
logging.basicConfig(level=logging.INFO)


class Some(Base):
    __tablename__ = 'some'

    id = sa.Column(sa.Integer, primary_key=True)
    some_value = sa.Column(sa.Text)


Base.metadata.create_all()

session = Session()

data = [dict(some_value='asd'), dict(some_value='qwe')]

logging.getLogger('sqlalchemy.engine').setLevel(logging.INFO)
session.bulk_save_objects([Some(**args) for args in data], return_defaults=True)
session.bulk_insert_mappings(Some, data, return_defaults=True)

print(sa.insert(Some).returning(Some.id).values(data))
```
Simple example with output:
```
INFO:sqlalchemy.engine.base.Engine:INSERT INTO "some" (some_value) VALUES (%(some_value)s) RETURNING "some".id
INFO:sqlalchemy.engine.base.Engine:{'some_value': 'asd'}
INFO:sqlalchemy.engine.base.Engine:INSERT INTO "some" (some_value) VALUES (%(some_value)s) RETURNING "some".id
INFO:sqlalchemy.engine.base.Engine:{'some_value': 'qwe'}
INFO:sqlalchemy.engine.base.Engine:INSERT INTO "some" (some_value) VALUES (%(some_value)s) RETURNING "some".id
INFO:sqlalchemy.engine.base.Engine:{'some_value': 'asd'}
INFO:sqlalchemy.engine.base.Engine:INSERT INTO "some" (some_value) VALUES (%(some_value)s) RETURNING "some".id
INFO:sqlalchemy.engine.base.Engine:{'some_value': 'qwe'}
```
And desirable query are:
```
INSERT INTO "some" (id, some_value) VALUES (%(id_m0)s, %(some_value_m0)s), (%(id_m1)s, %(some_value_m1)s) RETURNING "some".id
```

**Comment by zzzeek** [comment]
hi there -

psycopg2 does not support fetching of results when executemany() is used.  I haven't checked if this situation has changed recently, feel free to research this but it seems unlikely, as the default implementation for executemany() in psycopg2 runs the same statement over and over again, suggesting it would need to concatenate results together.

psycopg2 now supports extensions which SQLAlchemy also supports that will cause it to perform executemany() style batching differently, in that it can for an INSERT construct a single statement with the many sets of VALUES which you refer towards above, this is available via https://docs.sqlalchemy.org/en/13/dialects/postgresql.html#psycopg2-fast-execution-helpers .   The insert handler specifically uses a new psycopg2 method execute_batch().  This method does seem to support a "fetch" flag, however within SQLAlchemy this method is still invoked within the overall executemany() mechanics, which do not support fetching of results.

Overall no DBAPI executemany() method includes result fetching support so SQLAlchemy does not currently have the internal architecutral concepts to support this.  the psycopg2 execute_batch() method is the sole exception to where this would even be possible.    for SQLAlchemy to support this would require not just extremely major changes to Core but also to the ORM unit of work processes, which would involve significant developer resources.    It would be an unfortunately very large amount of new complexity to support a single use case on a single DBAPI but as PostgreSQL is extremely popular, I would support this effort to allow SQLAlchemy to support fetching of rows with executemany() but there is no plans on the roadmap now without highly motivated developers to take this on and see it through.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**Non working: ORM executemany returning** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1896

**Comment by zzzeek** [comment]
most of the work is in Core.  the above has rudimental changes for ORM however a bigger refactoring is proposed.

**Comment by zzzeek** [comment]
OK I figured this out, we can put a filter on the new Result object so that the ORM can fetch the returning rows with the "inserted primary key" and "returned defaults" right on top of it, so we wouldn't have to dig into the existing "inserted primary key" stuff.   also propose we flip psycopg2 to "insert_values" mode by default if the driver supports it.   can't know if it works until we just put it out there.

**Comment by zzzeek** [comment]
because yes, this would be an enormous speed improvement to the ORM on Postgresql.  like order of magnitude, for all ORM INSERT operations.

**Comment by kataev** [comment]
Very interesting, ty! If i undestand correctly, we have two paths to implement this: complicated with core reworking and relatively easy with orm + Result object. Your sample code for which path?
If i'm motivated for this change, from what i must to start? Can you recommend some failed test or work direction?

**Comment by zzzeek** [comment]
The Result object is a new thing that I've only begun to implement internally, so the state of current master is not ready for what I'm proposing here.     the ORM is currently on track to have very significant internal changes that I'm in the middle of right now.   so...I think right now the value of this issue is that being able to suit this kind of use case should be included within the work that I'm doing.

basically when you posted this issue I forgot that PG's execute_values() includes FETCH support.

**Comment by zzzeek** [comment]
the patch applies to the former approach.   I'd prefer to find a way that the persistence code doesn't have several huge conditionals as it does now which would involve more refactoring of how the result object delivers information about inserted primary keys and server defaults.

**Comment by kataev** [comment]
Oh, ok! I will be interested in waiting for your new developments.

**Comment by zzzeek** [comment]
OK so the patch at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1896 is working, and...thanks for the idea!  it's quite amazing.   it builds first on f1a3038f480ee1965928cdcd1dc0c47347f270bc for the Core feature, which was a bit of work, and the ORM part is in the above gerrit.

Here is an ORM insert of objects, without their primary key values so we have to fetch them back from SERIAL (or IDENTITY, upcoming feature) with RETURNING , so before the feature we have to run a separate insert for each of 100K objects (1.3):

$ python -m examples.performance bulk_inserts --dburl postgresql://scott:tiger@localhost/test 

`test_flush_no_pk : Individual INSERT statements via the ORM, calling upon last row id (100000 iterations); total time 16.173028 sec
`
Here is in 1.4 with the patch:

`test_flush_no_pk : Individual INSERT statements via the ORM, calling upon last row id (100000 iterations); total time 3.933317 sec
`

overall every variety of bulk INSERT is 3x-5x faster, whether or not we get the PK values back, because it's now executemany in all cases, and then if it is executemany, its a 5x faster executemany

this is the full suite for 1.3:

```
$ python -m examples.performance bulk_inserts --dburl postgresql://scott:tiger@localhost/test 
Tests to run: test_flush_no_pk, test_bulk_save_return_pks, test_flush_pk_given, test_bulk_save, test_bulk_insert_mappings, test_core_insert, test_dbapi_raw
test_flush_no_pk : Individual INSERT statements via the ORM, calling upon last row id (100000 iterations); total time 14.066174 sec
test_bulk_save_return_pks : Individual INSERT statements in "bulk", but calling upon last row id (100000 iterations); total time 14.643396 sec
test_flush_pk_given : Batched INSERT statements via the ORM, PKs already defined (100000 iterations); total time 7.925853 sec
test_bulk_save : Batched INSERT statements via the ORM in "bulk", discarding PKs. (100000 iterations); total time 6.748779 sec
test_bulk_insert_mappings : Batched INSERT statements via the ORM "bulk", using dictionaries. (100000 iterations); total time 5.241584 sec
test_core_insert : A single Core INSERT construct inserting mappings in bulk. (100000 iterations); total time 5.319086 sec
test_dbapi_raw : The DBAPI's API inserting rows in bulk. (100000 iterations); total time 6.078011 sec


```

then for 1.4 + the patch, the last entry is raw cursor.executemany(), which is amazingly slower than a full blown ORM insert, which is completely unprecedented.   The extreme slowness of psycopg2's default implementation here makes the whole thing very crazy:

```
$ python -m examples.performance bulk_inserts --dburl postgresql://scott:tiger@localhost/test
Tests to run: test_flush_no_pk, test_bulk_save_return_pks, test_flush_pk_given, test_bulk_save, test_bulk_insert_mappings, test_core_insert, test_dbapi_raw
test_flush_no_pk : Individual INSERT statements via the ORM, calling upon last row id (100000 iterations); total time 4.152247 sec
test_bulk_save_return_pks : Individual INSERT statements in "bulk", but calling upon last row id (100000 iterations); total time 3.465545 sec
test_flush_pk_given : Batched INSERT statements via the ORM, PKs already defined (100000 iterations); total time 3.987265 sec
test_bulk_save : Batched INSERT statements via the ORM in "bulk", discarding PKs. (100000 iterations); total time 2.529194 sec
test_bulk_insert_mappings : Batched INSERT statements via the ORM "bulk", using dictionaries. (100000 iterations); total time 1.206465 sec
test_core_insert : A single Core INSERT construct inserting mappings in bulk. (100000 iterations); total time 0.975962 sec
test_dbapi_raw : The DBAPI's API inserting rows in bulk. (100000 iterations); total time 5.659035 sec


```

**Comment by kataev** [comment]
ty, this is awesome!

**Comment by stevesmit** [comment]
@zzzeek Sorry if I've misunderstood...has this been patched into the current release 1.3.19? Or do I need to get 1.4 to take advantage of it. I also need the primary keys to be returned after a bulk insert (using Azure SQL server).

**Comment by zzzeek** [comment]
@stevesmit the upper right corner has the info on this issue: milestone is 1.4, so it's 1.4 only.  but also component is tagged as "postgresql".  this feature makes use of a psycopg2-specific API.    does not apply to pyodbc or pymssql right now, sorry.
